tools/validate_spec.py

```diff
@@ -79,81 +78,151 @@ def cli(argv: List[str]) -> Union[int, str]:
         return f"output file already exists: {args.json_output}"
 
     try:
-        validation_main(args)
+        identifier = ID(args.message_identifier)
+    except RecordFluxError as e:
+        return f'invalid identifier "{args.message_identifier}" : {e}'
+
+    try:
+        pdu_message = PyRFLX.from_specs(
+            [str(args.specification)], skip_model_verification=args.no_verification
+        )[str(identifier.parent)][str(identifier.name)]
+    except KeyError:
+        return f'message "{identifier.name}" could not be found in package "{identifier.parent}"'
+
+    except FileNotFoundError as e:
+        return f"specification {e}"
+
+    try:
+        validation_main(
+            pdu_message,
+            args.directory_invalid,
+            args.directory_valid,
+            args.json_output,
+            args.abort_on_error,
+        )
     except ValidationError as e:
         return f"{e}"
 
     return 0
 
 
-def validation_main(args: argparse.Namespace) -> None:
-    # pylint: disable=too-many-locals
-    def get_message_to_validate(message_path: Path, is_valid: bool) -> "OriginalMessage":
-        if not message_path.is_file():
-            raise ValidationError(f"{message_path} is not a regular file")
-        message_bytes = message_path.read_bytes()
-        message = OriginalMessage(message_bytes, is_valid, message_path.name)
-        return message
-
-    root_message_id: str = args.message_identifier
-    path_spec: Path = args.specification
-    path_invalid: Optional[Path] = args.directory_invalid
-    path_valid: Optional[Path] = args.directory_valid
-    full_output_path: Optional[Path] = args.json_output
-    abort_on_error: bool = args.abort_on_error
-    no_verification: bool = args.no_verification
-
-    path_generators: Dict[Path, Iterator[Path]] = {}
-    for path in [path_valid, path_invalid]:
-        if path is not None:
-            path_generators[path] = path.glob("*")
+def validation_main(
+    pdu_message: MessageValue,
+    directory_invalid: Optional[Path],
+    directory_valid: Optional[Path],
+    json_output: Optional[Path],
+    abort_on_error: bool,
+) -> None:
+    def validate_directory(path_iterator: Iterator[Path], is_valid_directory: bool) -> None:
+        for path in path_iterator:
+            validation_result = __validate_message(path, is_valid_directory, pdu_message)
+            output_writer.write_result(validation_result)
+            if not validation_result.validation_success and abort_on_error:
+                raise ValidationError(f"aborted: message {path} was classified incorrectly")
```

Comment:
If I change this to a private top-level function, `output_writer`, `pdu_message` and `abort_on_error` would need to be passed as arguments to it explicitly. I do not really like this since the values to those arguments never change.

The second approach would be this: 

```python3
valid_message_paths = directory_valid.glob("*") if directory_valid is not None else []
invalid_message_paths = directory_invalid.glob("*") if directory_invalid is not None else []

with OutputWriter(json_output) as output_writer:
    for path_iterator, is_valid_directory in [(valid_message_paths, True), (invalid_message_paths, False)]:
        for path in path_iterator:
            validation_result = __validate_message(path, is_valid_directory, pdu_message)
            output_writer.write_result(validation_result)
            if not validation_result.validation_success and abort_on_error:
                raise ValidationError(f"aborted: message {path} was classified incorrectly")
```

Do you think this is cleaner than using an inner function? I don`t really see the problem with using an inner function in this case, it results in a short statement and is explicit https://github.com/Componolit/RecordFlux-specifications/pull/44#discussion_r590221573 .

Reply:
The general problem of inner functions is that it is not obvious on which data the function depends. The worst part in the concrete case was the use of `output_writer` in the inner function before `output_writer` was actually defined.